### PR TITLE
Add Changeset.changed?/2,3

### LINKED ISSUE #4117: Add Ecto.Changeset.changed?(changeset, field, filters \\ [])
### Elixir version

Any

### Database and Version

Any

### Ecto Versions

All

### Database Adapter and Versions (postgrex, myxql, etc)

All

### Current behavior

No function exists.

### Expected behavior

I would like to add `Ecto.Changeset.changed?(changeset, field)` that returns a boolean if the field has changed. You can pass additional filters that will further restrict the verification. For example:

    Ecto.Changeset.changed?(changeset, :role, from: :admin)

will return true only if the field was previously an admin.

The following returns true only if the field was nilified:

    Ecto.Changeset.changed?(changeset, :role, to: nil)

Or only if the role changed from admin to editor:

    Ecto.Changeset.changed?(changeset, :role, from: :admin, to: :editor)

I feel this will make understanding how a field changed much more accessible. It doesn't cover some cases (the field was something BUT admin), but it should be generally useful.

### PR DESCRIPTION
**Pull request by v0idpwn**
Closes #4117

### CODE REVIEW

File: lib/ecto/changeset.ex [hunk 1]
```diff
@@ -454,6 +454,59 @@ defmodule Ecto.Changeset do
     end)
   end
 
+  @doc """
+  Returns true if a field was changed in a changeset
```
Comment (wojtekmach) [inline lib/ecto/changeset.ex:458]:
```suggestion
  Returns true if a field was changed in a changeset.
```

File: lib/ecto/changeset.ex [hunk 2]
```diff
@@ -454,6 +454,59 @@ defmodule Ecto.Changeset do
...
+      true
+
+      iex> changed?(changeset, :title, to: "NEW TITLE")
+      false
+  """
+  @spec changed?(t, atom, Keyword.t) :: boolean
+  def changed?(%Changeset{} = changeset, field, opts \\ []) when is_atom(field) do
+    case fetch_change(changeset, field) do
+      {:ok, new_value} ->
+        changed_from?(changeset, field, opts) and changed_to?(new_value, opts)
```
Comment (josevalim) [inline lib/ecto/changeset.ex:483]:
```suggestion
        Enum.all?(opts, fn
          {:from, from} -> Map.get(changeset.data, field) == from
          {:to, to} -> new_value == to
          other -> raise ArgumentError, "unknown option #{inspect(other)}"
        end)
```

**Comment by greg-rychlewski** [comment]
Sorry for the bad memory. Is it possible that the change associated with an association is a changeset? If so should we have special behaviour for associations or maybe even just raise?

File: lib/ecto/changeset.ex (hunk 2 above)
Comment (v0idpwn) [reply lib/ecto/changeset.ex:483]: That's clever. I was thinking that there should be a better way to do it :)

**Comment by josevalim** [comment]
@greg-rychlewski Good point about associations/embeds!

1. For relation == :one, we check if `action != :update or changes != %{}` 
2. For relation == :many, we check if at least one `action != :update or changes != %{}` 

However, we should raise if `:from` or `:to` are given to an association/embed.

File: lib/ecto/changeset.ex [hunk 3]
```diff
@@ -454,6 +454,43 @@ defmodule Ecto.Changeset do
...
+      iex> changed?(changeset, :title, to: "NEW TITLE")
+      false
+  """
+  @spec changed?(t, atom, Keyword.t) :: boolean
+  def changed?(%Changeset{} = changeset, field, opts \\ []) when is_atom(field) do
+    case fetch_change(changeset, field) do
+      {:ok, new_value} ->
+        Enum.all?(opts, fn
+          {:from, from} -> Map.get(changeset.data, field) == from
+          {:to, to} -> new_value == to
```
Comment (v0idpwn) [inline lib/ecto/changeset.ex:485]: Another note: should do semantic comparison for `:from` and `:to`

Comment (josevalim) [reply lib/ecto/changeset.ex:485]: Because of Decimal and stuff, right? Good call.

Comment (josevalim) [reply lib/ecto/changeset.ex:485]: @v0idpwn ping just in case you forgot. if you are busy, no worries, i would still wait 2-4 weeks for a new release.

**Comment by v0idpwn** [comment]
I pushed a new version that (I believe) handles all cases, but it hurts my aesthetic sense

**Review by josevalim — APPROVED** [review]
Beautiful!

Do we have any tests for Decimal or anything that uses Ecto.Type.equal? If so, ship it!

**Pull request merged**